### ISSUE TITLE
Hashrate Drop when using 2 threads behind PCIE Switch

### ISSUE THREAD

**Issue by Slippy369** [fireice-uk/xmr-stak#927]
Recently picked up one of these cheapo PCIE Switch devices:

https://www.amazon.com/dp/B077JNRFN2/ref=sspa_dk_detail_1?psc=1&pd_rd_i=B077JNRFN2&pd_rd_wg=sQt0N&pd_rd_r=N37QE95Y98RKM7JF0AW7&pd_rd_w=FVPcG

Everything seems fine all the GPU's behind it show up with no issues.  When I run Stak in single thread mode they work fine and I see ~750H/s which is to be expected when running a single thread per GPU.  Now when I double up each thread on the GPU's behind the switch things get weird.  Each GPU that is behind this device when running two threads ends up with less than half of the expected hashrate.  Getting about 200H/s per thread where normally when directly connected it would see 450H/s+ per thread.  I've tired different slots and each card individually so I don't think its a hardware issue.

Very well could be a driver issue but figured I would try here first.  Any ideas why the GPU's behind this PCI Bridge device would behave this way when running two threads.  I tried other miners (Cast) and saw similar bad results.  

My thought is that the extra level in the pcie hierarchy is causing the threads to conflict in someway?  When I run the cards behind the switch with only run thread everything seems stable but I'm leaving hashes on the table.  I started poking at the code but I'm not sure where to begin frankly

Running Windows 10 with 7 RX580's.  Latest Adrenalin drivers with compute mode enabled.  If there is any debug information that would be of help let me know and I'll get it as quickly as I can.

**Comment by kylemanna** [comment]
Curious about this as well. Certainly the PCIE switch is halving the bandwidth and introducing latency to each GPU, but I was under the impression that typically the CPU-GPU bandwidth and latency requirements were pretty low (relative to GPU gaming at least) for mining.

**Comment by Slippy369** [comment]
Agreed. I've seem videos of folks mining Ethereum based coins at full hash rate using these switches.  I might try that as an experiment at some point.  

See this video.  I have the same card this guy is using with no issues.
https://www.youtube.com/watch?v=ojFbLtZgddk

**Comment by dewminer** [comment]
Since I tested a lot of these PCIe packet switches (on Linux), what I can recommend You is:

- to use PCIe switches based on Pericom chip instead of ASmedia (more reliable)
- insert the packet switch in MB to PCIe slot connected to CPU (first x16 slot on MB or two x16 slots on intel Zxxx chipsets)

for example, the following shop offers 3 types of PCIe packet switches:
https://www.aliexpress.com/item/PCI-E-1-to-3-4-2-PCI-express-1X-slots-Riser-Card-Mini-ITX-to/32821608298.html

1to2 -> has asmedia asm1182e chip; without a heatsink it will burn out in few days
1to3 -> pericom PI7C9X2G; no issues at all
1to4 -> asmedia asm1184e; i had few of them in past, some ports where not working right from the beginning, burned out two of them in a weeks (even if a heatsink was there)

**Comment by Slippy369** [comment]
Thanks for the Information on these.  I work with PCIE switches regularly but typically its the 24-96 lane variety from other vendors.  I'm familiar with Pericom and expected that was the device I was going to find on the board.  When I looked up the vendor ID and could find nothing on asmedia I definitely wasn't thrilled.  Definitely will hunt down the Pericom variety for any future builds.

The motherboard I'm using has 3 x16 connectors from the CPU.  Its actually a server CPU with 40 pcie lanes so you end up with 2 full x16 connectors and one x8 connector all directly fed from the CPU.  I have tried this card in the x8 slot from the CPU and off one of the x1 connectors off the chipset and both behave the same way.  I will try a x16 slot when I get a chance just to see if that helps.

The weird thing is even if I only put 1 card behind this device it does the same thing.  Single threaded it runs up to 770h/s and when I run dual threaded it drops to 400h/s with all other settings being equal.  If I then remove the switch and plug the same card into that slot I get the 920h/s with two threads as expected.  I am currently running 3 cards behind the switch at 770, 750, and 700h/s for about 12 hours now.  All seems stable otherwise so I feel like there is some software aspect to this issue.

**Comment by JerichoJones** [comment]
@Slippy369 Please let us know what you find.
